File: backend/web-java/src/main/java/cn/yuanxin/mvp/web/idempotency/Jcs.java
package cn.yuanxin.mvp.web.idempotency;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;

/**
 * RFC 8785 (JCS) 序列化 + SHA-256 payload_hash（contracts/canonicalization.md）。
 *
 * <p><b>行为锁定基准是 backend/contracts/scripts/jcs.py</b>（跨语言唯一参考
 * 实现，两侧测试必须字节一致复现 samples/canonicalization/vectors.json 全部
 * 向量）。实现要点：</p>
 * <ul>
 *   <li>对象键按 UTF-16 code unit 升序（Java String.compareTo 即该顺序）；
 *       数组保持原序；无空白；</li>
 *   <li>字符串最小转义（" \ \b \t \n \f \r、&lt;0x20 → 反斜杠+u+4位小写hex），
 *       非 ASCII（中文等）原样 UTF-8 输出；孤立代理对拒绝；</li>
 *   <li>数字 ECMAScript Number::toString 风格；-0 → "0"；NaN/Infinity 拒绝；
 *       JSON 数字超过 2^53-1 拒绝（bigint 一律以字符串传输）；</li>
 *   <li>输入解析启用 Jackson STRICT_DUPLICATE_DETECTION——重复 JSON 键是
 *       错误而不是覆盖。</li>
 * </ul>
 *
 * <p>数字序列化为标准 ECMAScript {@code Number::toString} 语义
 * （RFC 8785 §4.2.2）：-0 → "0"；|v| &lt; 1e21 的整值 double → 无小数点的
 * 规范十进制；最短往返数字（JDK 19+ 的 {@link Double#toString(double)} 与
 * Python repr 同为最短往返表示，数字序列一致）；n 为小数点相对数字串的
 * 位置，剥离前导零后按 {@code n = intPart.length + exp10 - strippedZeros}
 * 修正（保证 0.5 → "0.5"、0.000123 → "0.000123" 与 jcs.py 逐字节一致）。</p>
 */
public final class Jcs {

    private static final BigInteger MAX_SAFE = BigInteger.valueOf(9007199254740991L);

    /** 严格 ObjectMapper：拒绝重复键、拒绝未知 token 之外的宽松行为。 */
    private static final ObjectMapper STRICT = new ObjectMapper()
            .enable(JsonParser.Feature.STRICT_DUPLICATE_DETECTION)
            .enable(DeserializationFeature.FAIL_ON_TRAILING_TOKENS);

    private Jcs() {
    }

    /** 解析 JSON 字符串；重复键/尾随内容/语法错误 → IllegalArgumentException。 */
    public static JsonNode parseStrict(String json) {
        try {
            JsonNode node = STRICT.readTree(json);
            if (node == null) {
                throw new IllegalArgumentException("empty JSON input");
            }
            return node;
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("invalid JSON input: " + e.getOriginalMessage(), e);
        }
    }

    /** 任意 Java 结构（Map/List/String/Integer/Long/Double/Boolean/null）→ JsonNode。 */
    public static JsonNode toNode(Object value) {
        JsonNode node = STRICT.valueToTree(value);
        if (node == null) {
            throw new IllegalArgumentException("value not convertible to JSON");
        }
        return node;
    }

    public static String canonicalize(JsonNode node) {
        StringBuilder sb = new StringBuilder();
        write(node, sb);
        return sb.toString();
    }

    public static byte[] canonicalizeBytes(JsonNode node) {
        return canonicalize(node).getBytes(StandardCharsets.UTF_8);
    }

    /** SHA-256(canonical UTF-8 字节)，小写 hex（无 "sha256:" 前缀；T13 列存裸 hex）。 */
    public static String sha256Hex(JsonNode node) {
        return sha256HexBytes(canonicalizeBytes(node));
    }

    public static String sha256HexBytes(byte[] bytes) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            return HexFormat.of().formatHex(md.digest(bytes));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    // ---------------- serialization core ----------------

    private static void write(JsonNode n, StringBuilder o) {
        switch (n.getNodeType()) {
            case NULL -> o.append("null");
            case BOOLEAN -> o.append(n.booleanValue() ? "true" : "false");
            case NUMBER -> o.append(numberToString(n));
            case STRING -> appendString(n.textValue(), o);
            case ARRAY -> {
                o.append('[');
                for (int i = 0; i < n.size(); i++) {
                    if (i > 0) o.append(',');
                    write(n.get(i), o);
                }
                o.append(']');
            }
            case OBJECT -> {
                List<String> names = new ArrayList<>();
                n.fieldNames().forEachRemaining(names::add);
                names.sort(String::compareTo); // UTF-16 code-unit order (ES6 sort)
                o.append('{');
                for (int i = 0; i < names.size(); i++) {
                    if (i > 0) o.append(',');
                    appendString(names.get(i), o);
                    o.append(':');
                    write(n.get(names.get(i)), o);
                }
                o.append('}');
            }
            default -> throw new IllegalArgumentException(
                    "unsupported JSON value type: " + n.getNodeType());
        }
    }

    private static void appendString(String s, StringBuilder o) {
        o.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> o.append("\\\"");
                case '\\' -> o.append("\\\\");
                case '\b' -> o.append("\\b");
                case '\t' -> o.append("\\t");
                case '\n' -> o.append("\\n");
                case '\f' -> o.append("\\f");
                case '\r' -> o.append("\\r");
                default -> {
                    if (Character.isHighSurrogate(c)) {
                        if (i + 1 >= s.length() || !Character.isLowSurrogate(s.charAt(i + 1))) {
                            throw new IllegalArgumentException(
                                    "lone surrogate U+" + Integer.toHexString(c) + " in string");
                        }
                        o.append(c).append(s.charAt(++i));
                    } else if (Character.isLowSurrogate(c)) {
                        throw new IllegalArgumentException(
                                "lone surrogate U+" + Integer.toHexString(c) + " in string");
                    } else if (c < 0x20) {
                        o.append("\\u").append(String.format("%04x", (int) c));
                    } else {
                        o.append(c); // 非 ASCII（含中文）原样输出
                    }
                }
            }
        }
        o.append('"');
    }

    /** 与 jcs.py::_number_to_string 一致的标准 ES6 Number::toString（RFC 8785 §4.2.2）。 */
    static String numberToString(JsonNode n) {
        if (n.isIntegralNumber()) {
            BigInteger v = n.bigIntegerValue();
            if (v.abs().compareTo(MAX_SAFE) > 0) {
                throw new IllegalArgumentException(
                        "JSON number exceeds the safe double range; transport big integers as strings");
            }
            return v.toString();
        }
        if (n.isBigDecimal()) {
            // 本项目 JSON 不产生 BigDecimal 数字节点；按 double 规则拒绝越界值
            java.math.BigDecimal bd = n.decimalValue();
            if (bd.abs().compareTo(new java.math.BigDecimal(MAX_SAFE)) > 0) {
                throw new IllegalArgumentException("JSON number exceeds the safe double range");
            }
            return doubleToString(bd.doubleValue());
        }
        if (!n.isFloatingPointNumber()) {
            throw new IllegalArgumentException("unsupported number node: " + n.getNodeType());
        }
        double d = n.doubleValue();
        return doubleToString(d);
    }

    static String doubleToString(double d) {
        if (Double.isNaN(d) || Double.isInfinite(d)) {
            throw new IllegalArgumentException("NaN/Infinity are not serializable per RFC 8785");
        }
        if (d == 0.0) {
            return "0"; // 覆盖 -0
        }
        String sign = d < 0 ? "-" : "";
        double a = Math.abs(d);
        if (a == Math.rint(a) && a < 1e21) {
            return sign + java.math.BigDecimal.valueOf(a).toBigIntegerExact().toString();
        }
        // 最短往返十进制（JDK 19+ Double.toString 与 Python repr 同为最短性+最近值，数字序列一致）
        String s = Double.toString(a);
        int eIdx = s.indexOf('E');
        int exp10 = 0;
        if (eIdx >= 0) {
            exp10 = Integer.parseInt(s.substring(eIdx + 1));
            s = s.substring(0, eIdx);
        }
        int dot = s.indexOf('.');
        String intPart = dot < 0 ? s : s.substring(0, dot);
        String fracPart = dot < 0 ? "" : s.substring(dot + 1);
        String raw = intPart + fracPart;
        String digits = raw.replaceFirst("^0+", "");
        if (digits.isEmpty()) digits = "0";
        // value == 0.raw * 10**(intPart.length()+exp10)；剥 z 个前导零后小数点位置左移 z
        int nPos = intPart.length() + exp10 - (raw.length() - digits.length());
        digits = digits.replaceFirst("0+$", "");
        if (digits.isEmpty()) digits = "0";
        int k = digits.length();
        if (k <= nPos && nPos <= 21) {
            return sign + digits + "0".repeat(nPos - k);
        }
        if (0 < nPos && nPos <= 21) {
            return sign + digits.substring(0, nPos) + "." + digits.substring(nPos);
        }
        if (-6 < nPos && nPos <= 0) {
            return sign + "0." + "0".repeat(-nPos) + digits;
        }
        String m = (k == 1) ? digits : digits.charAt(0) + "." + digits.substring(1);
        int e = nPos - 1;
        return sign + m + (e >= 0 ? "e+" : "e-") + Math.abs(e);
    }

    /** 便捷：从字段顺序无关的 Map 构造 canonical SHA-256。 */
    public static String sha256HexOfObject(Object javaStructure) {
        JsonNode root = toNode(javaStructure);
        if (!root.isObject()) {
            throw new IllegalArgumentException("canonical object must be a JSON object");
        }
        return sha256Hex(root);
    }

    /** 单元测试辅助：返回 canonical ObjectNode（键序无关）。 */
    public static ObjectNode objectNode() {
        return STRICT.createObjectNode();
    }

    public static ArrayNode arrayNode() {
        return STRICT.createArrayNode();
    }
}

File: backend/contracts/scripts/jcs.py
#!/usr/bin/env python3
"""Dependency-free RFC 8785 (JSON Canonicalization Scheme) reference implementation.

Part of the MVP foundation contract (backend/contracts). Both the Java web
service and the Python worker must reproduce this canonicalization for
T13 payload_hash and cross-language contract tests; see
backend/contracts/canonicalization.md for the request-object spec and
backend/contracts/samples/canonicalization/vectors.json for shared vectors.

CLI:
    python3 scripts/jcs.py hash <file.json>      # print lowercase sha256 hex
    python3 scripts/jcs.py serialize <file.json> # print canonical UTF-8 bytes
    python3 scripts/jcs.py selftest              # ES6 Number::toString pairs self-check

Rules implemented (RFC 8785):
  * object members sorted by key in UTF-16 code-unit order (ES6 sort),
    serialized with no whitespace;
  * arrays keep their order (order-significant);
  * strings serialized as minimal-escape JSON strings, output UTF-8 (U+FEFF
    and other non-ASCII characters are NOT escaped);
  * numbers serialized per ECMAScript Number::toString (finite only;
    -0 -> "0"; integral doubles -> canonical decimal without leading zeros;
    exponent form "e+XX"/"e-XX" without leading zeros in the exponent);
  * duplicate keys in the input JSON object are rejected (JCS requires
    sequences of unique names);
  * lone surrogates in keys/values are rejected: they cannot be encoded as
    valid UTF-8 and never appear in our contract data.
"""

from __future__ import annotations

import hashlib
import json
import math
import sys

MAX_SAFE_INTEGER = 9007199254740991  # 2**53 - 1


class CanonicalizationError(ValueError):
    """Raised when input cannot be canonicalized per RFC 8785."""


def _utf16_code_units(s: str) -> list[int]:
    units: list[int] = []
    for ch in s:
        cp = ord(ch)
        if 0xD800 <= cp <= 0xDFFF:
            raise CanonicalizationError(
                f"lone surrogate U+{cp:04X} in string {s!r}; not representable"
            )
        if cp > 0xFFFF:
            cp -= 0x10000
            units.append(0xD800 + (cp >> 10))
            units.append(0xDC00 + (cp & 0x3FF))
        else:
            units.append(cp)
    return units


def _sort_key(s: str) -> list[int]:
    return _utf16_code_units(s)


_SHORT_ESCAPES = {
    '"': '\\"',
    "\\": "\\\\",
    "\b": "\\b",
    "\f": "\\f",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
}


def _serialize_string(s: str) -> str:
    out = ['"']
    for ch in s:
        esc = _SHORT_ESCAPES.get(ch)
        if esc is not None:
            out.append(esc)
            continue
        cp = ord(ch)
        if 0xD800 <= cp <= 0xDFFF:
            raise CanonicalizationError(
                f"lone surrogate U+{cp:04X} in string {s!r}; not representable"
            )
        if cp < 0x20:
            out.append("\\u%04x" % cp)
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _number_to_string(value: float | int) -> str:
    """ECMAScript Number::toString(value) for finite doubles (radix 10).

    Python ``int`` values outside the exact-double range are rejected: JSON
    numbers in our contracts are either small integers (schema_version) or
    arrive as bigint-as-string, never as > 2**53 JSON numbers.
    """
    if isinstance(value, int):
        if abs(value) > MAX_SAFE_INTEGER:
            raise CanonicalizationError(
                f"JSON number {value} exceeds the safe double range; "
                "transport big integers as strings"
            )
        return str(value)
    f = value
    if math.isnan(f) or math.isinf(f):
        raise CanonicalizationError("NaN/Infinity are not serializable per RFC 8785")
    if f == 0.0:
        return "0"  # covers -0 as well
    d = float(f)
    if d.is_integer() and abs(d) < 1e21:
        return str(int(d))
    # Shortest round-trip decimal digits (Python repr == JS DoubleToRawPrecision).
    a = abs(d)
    s = repr(a)
    if "e" in s:
        mant, exp = s.split("e")
        exp = int(exp)
    else:
        mant, exp = s, 0
    int_part, _, frac_part = mant.partition(".")
    raw = int_part + frac_part
    digits = raw.lstrip("0") or "0"
    # n: position of the decimal point relative to the digit string,
    # i.e. value == 0.digits * 10**n. Before stripping, value ==
    # 0.raw * 10**(len(int_part) + exp); dropping z leading zeros from the
    # digit string shifts the decimal point z places left (fixes |v| < 1,
    # where int_part is "0" — e.g. 0.5 -> n=0, not n=1).
    n = len(int_part) + exp - (len(raw) - len(digits))
    # Strip trailing zeros from digits, adjusting nothing about n (they were
    # never trailing in the integer sense).
    digits = digits.rstrip("0") or "0"
    k = len(digits)
    sign = "-" if d < 0 else ""
    if k <= n <= 21:
        return sign + digits + "0" * (n - k)
    if 0 < n <= 21:
        return sign + digits[:n] + "." + digits[n:]
    if -6 < n <= 0:
        return sign + "0." + "0" * (-n) + digits
    if k == 1:
        m = digits
    else:
        m = digits[0] + "." + digits[1:]
    e = n - 1
    return sign + m + ("e+" if e >= 0 else "e-") + str(abs(e))


# ECMAScript Number::toString pairs verified against the ES6 spec (authoritative
# for RFC 8785 §4.2.2). 9007199254740992 is the *double* 2**53 (integer literals
# above 2**53-1 are rejected by this contract: bigint-as-string, see docs).
_NUMBER_SELFTEST_PAIRS: list[tuple[float, str]] = [
    (0.0, "0"),
    (-0.0, "0"),
    (1.0, "1"),
    (0.5, "0.5"),
    (0.000123, "0.000123"),
    (1e-4, "0.0001"),
    (1e-5, "0.00001"),
    (1e-6, "0.000001"),
    (1e-7, "1e-7"),
    (1e20, "100000000000000000000"),
    (1e21, "1e+21"),
    (1e23, "1e+23"),
    (1e30, "1e+30"),
    (3.14159, "3.14159"),
    (123.456, "123.456"),
    (9007199254740992.0, "9007199254740992"),
    (1.7976931348623157e308, "1.7976931348623157e+308"),
    (2.2250738585072014e-308, "2.2250738585072014e-308"),
    (0.1, "0.1"),
    (-0.5, "-0.5"),
    (1e-3, "0.001"),
    (2.5e-8, "2.5e-8"),
    (1234.5, "1234.5"),
]


def selftest() -> int:
    """Assert ES6 Number::toString pairs and a few structural rules. 0 = pass."""
    failures = 0
    for value, expected in _NUMBER_SELFTEST_PAIRS:
        got = _number_to_string(value)
        if got != expected:
            failures += 1
            print(f"FAIL: {value!r} -> {got!r}, expected {expected!r}")
    checks = [
        (canonicalize({"b": 1, "a": [2, {"d": None, "c": True}]}),
         '{"a":[2,{"c":true,"d":null}],"b":1}'),
        (canonicalize({"成员": "张三", "A": 1}), '{"A":1,"成员":"张三"}'),
    ]
    for got, expected in checks:
        if got != expected:
            failures += 1
            print(f"FAIL: structural {got!r} != {expected!r}")
    try:
        load_strict('{"a":1,"a":2}')
        failures += 1
        print("FAIL: duplicate keys must be rejected")
    except CanonicalizationError:
        pass
    total = len(_NUMBER_SELFTEST_PAIRS) + len(checks) + 1
    if failures:
        print(f"selftest: {failures}/{total} FAILED")
        return 1
    print(f"selftest: PASS ({total} checks, {len(_NUMBER_SELFTEST_PAIRS)} number pairs)")
    return 0


def _reject_duplicate_pairs(pairs: list[tuple[str, object]]) -> dict:
    seen: set[str] = set()
    for key, _ in pairs:
        if key in seen:
            raise CanonicalizationError(f"duplicate JSON object key {key!r}")
        seen.add(key)
    return dict(pairs)


def canonicalize(value: object) -> str:
    """Return the RFC 8785 canonical serialization of a parsed JSON value."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return _serialize_string(value)
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return _number_to_string(value)
    if isinstance(value, list):
        return "[" + ",".join(canonicalize(item) for item in value) + "]"
    if isinstance(value, dict):
        parts = []
        for key in sorted(value.keys(), key=_sort_key):
            parts.append(_serialize_string(key) + ":" + canonicalize(value[key]))
        return "{" + ",".join(parts) + "}"
    raise CanonicalizationError(f"unsupported JSON value type: {type(value).__name__}")


def canonicalize_bytes(value: object) -> bytes:
    return canonicalize(value).encode("utf-8")


def sha256_hex(value: object) -> str:
    return hashlib.sha256(canonicalize_bytes(value)).hexdigest()


def load_strict(text: str) -> object:
    """Parse JSON rejecting duplicate object keys (JCS precondition)."""
    return json.loads(text, object_pairs_hook=_reject_duplicate_pairs)


def _usage() -> "None":
    prog = sys.argv[0]
    print(f"usage: python3 {prog} hash <file.json>", file=sys.stderr)
    print(f"       python3 {prog} serialize <file.json>", file=sys.stderr)
    print(f"       python3 {prog} selftest", file=sys.stderr)


def main(argv: list[str]) -> int:
    if argv == ["selftest"]:
        return selftest()
    if len(argv) != 2 or argv[0] not in ("hash", "serialize"):
        _usage()
        return 2
    mode, path = argv[0], argv[1]
    try:
        with open(path, "r", encoding="utf-8") as fh:
            value = load_strict(fh.read())
        if mode == "hash":
            print(sha256_hex(value))
        else:
            sys.stdout.buffer.write(canonicalize_bytes(value) + b"\n")
        return 0
    except (OSError, json.JSONDecodeError, CanonicalizationError) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
